### PR TITLE
Fix TransactionStatusMeta breakage in blockstore

### PR DESCRIPTION
**Pull request by CriesofCarrots**
#### Problem
As of https://github.com/solana-labs/solana/pull/12311 (and the v1.3 backport), nodes running on current code cannot deserialize legacy TransactionStatusMeta data.

#### Summary of Changes
- Add a deserialization helper function that sets remaining fields to default on end of file error. This enables extending structs by adding new optional fields to the end.

cc #12494

### CODE REVIEW

File: sdk/src/deserialize_utils.rs [hunk 1]
```diff
@@ -0,0 +1,91 @@
+use serde::{Deserialize, Deserializer};
+
+pub fn default_on_eof<'de, T, D>(d: D) -> Result<T, D::Error>
+where
+    D: Deserializer<'de>,
+    T: Deserialize<'de> + Default,
+{
+    let result = T::deserialize(d);
+    match result {
+        Err(err) if err.to_string() == "io error: unexpected end of file" => Ok(T::default()),
```
Comment (CriesofCarrots) [inline sdk/src/deserialize_utils.rs:10]: @mvines I spent a bit of time trying to check the error type here, but didn't make a lot of headway, due to the boxed generic error that serde deserialize returns (to accommodate different formats, etc). Got any ideas?

Comment (jstarry) [reply sdk/src/deserialize_utils.rs:10]: I also spent time there and didn't make any progress, I think this is the best way

Comment (mvines) [reply sdk/src/deserialize_utils.rs:10]: ok, wfm

File: sdk/src/deserialize_utils.rs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    use bincode::deserialize;
+
+    #[test]
+    fn test_default_on_eof() {
+        #[derive(Serialize, Deserialize, Debug, PartialEq)]
+        struct Foo {
+            bar: u16,
+            #[serde(deserialize_with = "default_on_eof")]
+            baz: Option<u16>,
+            #[serde(deserialize_with = "default_on_eof")]
```
Comment (mvines) [inline sdk/src/deserialize_utils.rs:27]: I'm curious what would happen if this second `default_on_eof` was omitted?   Does `deserialize` blow up?

Comment (CriesofCarrots) [reply sdk/src/deserialize_utils.rs:27]: If the second `default_on_eof` is omitted: if `bar` and `baz` are both present, `deserialize` works fine, but if `deserialize` on a partial struct blows up.

Comment (CriesofCarrots) [reply sdk/src/deserialize_utils.rs:27]: I can add a should-panic unit test to emphasize usage, if that would be helpful...?

Comment (mvines) [reply sdk/src/deserialize_utils.rs:27]: Yeah a test demoing the blow-up that we can point people at when their code blows up seems helpful.  Sucks that we can't have serde attributes depend on each other

**Review by mvines — DISMISSED** [review]

**Pull request merged**